### ISSUE TITLE
Scintilla5: text is mirrored in RTL

### ISSUE THREAD

**Issue by Yaron10** [notepad-plus-plus/notepad-plus-plus#11476]
**STR:**
Paste `Hello`.
Change direction to RTL.

**Result:**
The text is mirrored.

![תמונה](https://user-images.githubusercontent.com/10229320/161799708-2eb30ef5-63bf-4cc5-914a-762d9ec3d53a.png)

***
Notepad++ v8.3.3   (64-bit)
Build time : Apr  5 2022 - 15:55:23
Path : C:\---\Notepad++.x64.Release.exe
Command Line : 
Admin mode : ON
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 10 Pro (64-bit) 
OS Version : 2009
OS Build : 19042.985
Current ANSI codepage : 1255
Plugins : ComparePlus.dll HTMLTag.dll MenuIcons.dll PythonScript.dll _CustomizeToolbar.dll

**Comment by Yaron10** [comment]
https://sourceforge.net/p/scintilla/feature-requests/1435/.

This is a **major issue** for RTL users.
Fixing it would be highly appreciated.

**Comment by donho** [comment]
@Yaron10 
It seems to me that RTL & LTR are broken in v8.3.3. Is it not the case to you?

**Comment by rdipardo** [comment]
RTL on Windows is described as "experimental" in [Scintilla's documentation][0]; where it also says:

> Only UTF-8 documents will show bidirectional behaviour and only in `SC_BIDIRECTIONAL_L2R` mode. Some features, such as [virtual space][1] may not work with bidirectional text or may work only in some circumstances. `SC_BIDIRECTIONAL_R2L` may be implemented in the future.

Probably not that help to mention here, but Scintilla <= 5.2.1 will [reportedly][-1] flip punctuation characters on GTK platforms (where RTL is [not officially supported][2]; users have to rely on system locale instead). Because of faulty character indexing, a locale like `LANG=he_IL.UTF-8` would cause a fatal access violation, before being [patched in 5.2.2][3] &#x2014; for GTK only. No idea if something similar can be ported to Win32.


[-1]: https://sourceforge.net/p/scintilla/bugs/2309
[0]: https://www.scintilla.org/ScintillaDoc.html#SCI_SETBIDIRECTIONAL
[1]: https://github.com/notepad-plus-plus/notepad-plus-plus/commit/88ed851478b3426e9cca365b67a1ceb4cd4f92e2
[2]: https://github.com/geany/geany/issues/3101#issuecomment-1013984738#:~:text=Scintilla%20does%20not%20deliberately%20support%20any%20directionality%20other%20than%20LTR%20on%20the%20GTK%20backend


[3]: https://sourceforge.net/p/scintilla/code/ci/0759b3980fda03919b33d7be85a1653d27b8437f/

**Comment by Yaron10** [comment]
@rdipardo,

Thank you for looking into it. 👍 

> RTL on Windows is described as "experimental"...

Yes, there were some issues prior to the recent Scintilla update.
But now the text (in Latin or any RTL language) is literally unreadable if the direction is RTL.

**Comment by Yaron10** [comment]
@rdipardo,

Just wanted to add that you can test it Spanish.
"View -> Change Text Direction to RTL".

Thanks again.

**Comment by donho** [comment]
@Yaron10 
> What do you mean by "broken"?

After running RTL command:

![image](https://user-images.githubusercontent.com/90293/162306222-ee6a93ba-fed2-40be-ac7f-08202b0e613b.png)

I guess it's rather a Windows 11 issue.

**Comment by donho** [comment]
> I guess it's rather a Windows 11 issue.

But on the same machine, v8.3 works as expected.

**Comment by rdipardo** [comment]
### Conclusion 

N++ is calling into a [Win32 API][0] when it should be using [Scintilla's API][1]:

> Scintilla has never actively supported this and any past success with these modes was accidental. May have worked with GDI drawing but not DirectWrite.  <https://sourceforge.net/p/scintilla/feature-requests/1435/#be47>

[0]: https://github.com/notepad-plus-plus/notepad-plus-plus/blob/7fa6fb083b0ab454be4a1565b8747732cdea9cc4/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp#L3668-L3673
[1]: https://www.scintilla.org/ScintillaDoc.html#SCI_SETBIDIRECTIONAL

**Comment by rdipardo** [comment]
== not a Scintilla bug

**Comment by Yaron10** [comment]
@donho,

What a catch! 👍 
(I should have found it :) ).

It happens when **DirectWrite is enabled** and I can reproduce it in older versions as well.
Did you have DirectWrite disabled when testing v8.3?
([Neil Hodgson's reply](https://sourceforge.net/p/scintilla/feature-requests/1435/#be47), quoted in @rdipardo's previous post here).

Before the recent Scintilla update, we could have said that RTL is not supported when DirectWrite is enabled.
But now the recommended `SCI_SETBIDIRECTIONAL` use seems necessary. 

Would you have time to look into it? - I'd be grateful.
(With my coding level, it might take me a long time. Sorry.).

Thank you.

***
@rdipardo,

Thank you for the important additional info. 👍

**Comment by donho** [comment]
@Yaron10 & @rdipardo 
Thank you for your information.

After modifying RTL method by using Scintilla Bi-directional API as such, with UTF8 document and DirectWrite enabled:
```
void ScintillaEditView::changeTextDirection(bool isRTL)
{
	if (isRTL)
	{
		execute(SCI_SETBIDIRECTIONAL, SC_BIDIRECTIONAL_R2L);
	}
	else
	{
		execute(SCI_SETBIDIRECTIONAL, SC_BIDIRECTIONAL_L2R);
	}
}
```

RTL doesn't work.

After reading the doc in the official site regarding the Bi-directional support, I'm totally confused:

SCI_GETBIDIRECTIONAL → int
Some languages, like Arabic and Hebrew, are written from right to left instead of from left to right as English is. Documents that use multiple languages may contain both directions and this is termed "bidirectional". The default text direction may be right to left or left to right. Scintilla only correctly displays bidirectional text on some platforms. Currently, there is experimental support for bidirectional text on Win32 using DirectWrite and on macOS using Cocoa. Only UTF-8 documents will show bidirectional behaviour and **only in SC_BIDIRECTIONAL_L2R mode**. Some features, such as virtual space may not work with bidirectional text or may work only in some circumstances. **SC_BIDIRECTIONAL_R2L may be implemented in the future.**

So is R2L of Scintilla not implemented yet? What's the purpose of this API if RTL is not working?

**Comment by nyamatongwe** [comment]
> @donho: So is R2L of Scintilla not implemented yet?

R2L mode was partly implemented but problems were found so that part wasn't committed. The API was retained in case further progress was made but the original implementer stopped work. The primary scenario for `SC_BIDIRECTIONAL_L2R` is source code in Roman (L2R) characters with islands of Arabic or Hebrew (R2L) in string literals and comments.

Most of the bidirectional development occurred in an email thread where links to the R2L code can be found if anyone wants to continue work.
https://groups.google.com/g/scintilla-interest/c/6u5YFzkRzRE/m/hdW8WY2XAQAJ

**Comment by rdipardo** [comment]
Did some experimenting with [Noteapd2] and [Notepad3]. 
In both, RTL will only take effect when GDI rendering is selected, and the results are equally illegible.

#### Notepad2 4.22.03 r4130 (4ddfdb81)
1. Settings > Advanced Settings > Rendering Technology > "Legacy GDI" or "DirectWrite GDI DC"
2. Settings > Advanced Settings > RTL Layout (GDI) > Edit Window

*Notes*
* the "Edit Window" option will be inactive _unless_ GDI rendering is active (a big hint that DirectWrite is known to be unsupported)
* the "Bidirectional (Direct2D)" menu options do nothing &#x2014; a known issue [^1]

![np2_gdi_rtl](https://user-images.githubusercontent.com/59004801/162553267-2743e660-9741-40f5-ae2a-e4264df1e1a5.png)

#### Notepad3 v5.21.1129.1 (ba6de2c2)
1. Settings > Technology > Rendering > "GDI (no ligatures)" or "DirectWrite GDI DC" 
2. Settings > Technology > RTL Layout (GDI) > Edit Area

*Note*
* once again the "Bidirectional" menu options do nothing &#x2014; another known issue: https://github.com/rizonesoft/Notepad3/issues/2639

![np3_gdi_rtl](https://user-images.githubusercontent.com/59004801/162553220-967556b1-b4ed-4d1c-a700-909774ac40fd.png)


As you can see above, characters and even line numbers are rendered backwards.
I expect this behaviour to be reproducible across Windows versions, so may be something worth studying.

[^1]: Notepad2 does make use of [the `SC_BIDIRECTIONAL_R2L` message][0], to no actual effect: https://github.com/zufuliu/notepad2/issues/392

[0]: https://github.com/zufuliu/notepad2/blob/1c47abf43ef79fa1d80bbc89b5239fe90fa9f6bb/src/Notepad2.c#L5645
[1]: https://github.com/zufuliu/notepad2/issues/392
[Noteapd2]: https://github.com/zufuliu/notepad2/issues/392
[Notepad3]: https://github.com/rizonesoft/Notepad3/issues/2639

**Comment by donho** [comment]
@nyamatongwe 
Thank you Neil for the explanation. I understand better the context now.
Since the implementation of bi-directional feature has stopped, and the mirrored text (caused by WS_EX_LAYOUTRTL) happens in Scintilla5, is there any chance that Scintilla5 restores the way to manage the text direction in Scintilla4 when application uses WS_EX_LAYOUTRTL ?
In this way, the Arabic or Hebrew users can continue using this feature.

**Comment by nyamatongwe** [comment]
@donho 
Scintilla 5 did not disable `WS_EX_LAYOUTRTL`. Any right-to-left behaviour with GDI and `WS_EX_LAYOUTRTL` in Scintilla 4 was fortuitous. It was never a planned feature and no work was done to implement it or remove it.  Comparing Notepad++ 8.2.1 (older Scintilla and selecting RTL) with current SciTE with Scintilla 5 plus a hardcoded `WS_EX_LAYOUTRTL` shows GDI still behaves the same. 

In this mode styled left-to-right code is drawn as a sequence of lexemes with the characters in each lexeme in left-to-right order but with the lexemes in right-to-left order. It may appear something like:

                                                               ;(hInstance)<HINSTANCE>static_cast = hinstPlatformRes

Is this what the Arabic and Hebrew users want? Or do they just want to edit unstyled prose files?

Implementing a similar appearance to `GDI+WS_EX_LAYOUTRTL` with DirectWrite may be possible but it is unclear to me just how to do that.

**Comment by Yaron10** [comment]
@donho, @nyamatongwe & @rdipardo,

Thank you very much for the thorough testing and analyzing. I appreciate it.  

> is there any chance that Scintilla5 restores the way to manage the text direction in Scintilla4 when application uses WS_EX_LAYOUTRTL ?

This is the question I would have asked at this point.

> In this mode styled left-to-right code is drawn as a sequence of lexemes with the characters in each lexeme in left-to-right order but with the lexemes in right-to-left order. It may appear something like:
```
                                                           ;(hInstance)<HINSTANCE>static_cast = hinstPlatformRes
```

This is the result in the best-RTL-support editors.
Please test it with `Hello`.

> Or do they just want to edit unstyled prose files?

This is the key for understanding the issue.
1. I want to be able to read and edit Hebrew "normal text" files. 
2. I want to be able to read and edit source code in Roman (L2R) characters with islands of Hebrew (R2L) in string literals and comments.

> Comparing Notepad++ 8.2.1 (older Scintilla and selecting RTL) with current SciTE with Scintilla 5 plus a hardcoded WS_EX_LAYOUTRTL shows GDI still behaves the same.

At this point the requested questions are:
1. What can possibly cause the current behavior in Notepad++ with Scintilla 5?
2. I suppose SciTE does not use something similar to:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/7fa6fb083b0ab454be4a1565b8747732cdea9cc4/PowerEditor/src/NppCommands.cpp#L3529
Not having to use that hack would be a nice bonus.

> Implementing a similar appearance to GDI+WS_EX_LAYOUTRTL with DirectWrite may be possible

That would be great. But let's please solve the current GDI issue first.

**Comment by nyamatongwe** [comment]
> @Yaron10  Please test it with Hello

![HelloRTL](https://user-images.githubusercontent.com/225818/162573897-3592ec8d-baec-4d5c-b5fc-32ef94d14386.png)

> What can possibly cause the current behavior in Notepad++ with Scintilla 5?

Inverted character glyphs appear to occur with `SC_TECHNOLOGY_DIRECTWRITEDC` and not the other `SC_TECHNOLOGY_DIRECTWRITE*`. DirectWrite may be drawing right way round but the transform to the DC is inverted horizontally.

> I suppose SciTE does not use something similar to
> // Wrap then !wrap to fix problem of mirror characters

No. I would guess that Notepad++ added that to try to clear cached layouts which have direction baked in. Its switching direction on the window but Scintilla doesn't get told that has occurred and has no way of handling it.

> But let's please solve the current GDI issue first.

Do the horizontally inverted glyphs (the initial illustration) occur with GDI?

**Comment by Yaron10** [comment]
@nyamatongwe,

> What can possibly cause the current behavior in Notepad++ with Scintilla 5?
>> Inverted character glyphs appear to occur with SC_TECHNOLOGY_DIRECTWRITEDC and not the other SC_TECHNOLOGY_DIRECTWRITE*. DirectWrite may be drawing right way round but the transform to the DC is inverted horizontally.

If I understand you correctly, you're referring to DirectWrite.
My question was about **GDI**: correct with Scintilla 4 vs. mirrored with Scintilla 5.

> I suppose SciTE does not use something similar to
> // Wrap then !wrap to fix problem of mirror characters
>> No. I would guess that Notepad++ added that to try to clear cached layouts which have direction baked in. Its switching direction on the window but Scintilla doesn't get told that has occurred and has no way of handling it.

Without that hack, the text would be mirrored if `Word wrap` was ON.
I assume it's not mirrored in your tests even if `Word wrap` is ON.

> Do the horizontally inverted glyphs (the initial illustration) occur with GDI?

I don't like ClearType and I've modified the fonts in Windows 10 Registry.
I get a "raw" font with GDI.
![תמונה](https://user-images.githubusercontent.com/10229320/162577244-c5d90a0a-77f1-4d0d-ad64-e724f71a11bd.png)

Could someone else test it please?

But anyway, that would be a relatively minor issue.

Thanks again.

**Comment by donho** [comment]
> No. I would guess that Notepad++ added that to try to clear cached layouts which have direction baked in. Its switching direction on the window but Scintilla doesn't get told that has occurred and has no way of handling it.

Yes, indeed.


>>>  What can possibly cause the current behavior in Notepad++ with Scintilla 5?

>>  Inverted character glyphs appear to occur with SC_TECHNOLOGY_DIRECTWRITEDC and not the other SC_TECHNOLOGY_DIRECTWRITE*. DirectWrite may be drawing right way round but the transform to the DC is inverted horizontally.

> If I understand you correctly, you're referring to DirectWrite.
My question was about GDI: correct with Scintilla 4 vs. mirrored with Scintilla 5.

@nyamatongwe Neil, I confirm that only GUI (*WS_EX_LAYOUTRTL*, without *SCI_GETBIDIRECTIONAL*) is used in the current implementation of Notepad++, and it works without enabling *SC_TECHNOLOGY_DIRECTWRITE*. This behaviour's been changed after updating Scintilla from v4.4.6 (RTL: `Hello`) to v5.2.1 (RTL: `olleH`) and *SC_TECHNOLOGY_DIRECTWRITE*,  *SC_TECHNOLOGY_DIRECTWRITERETAIN* and *SC_TECHNOLOGY_DIRECTWRITE* are not enabled.

That makes us think the text direction management might have been changed in Scintilla5 with Microsoft legacy GDI (by setting *WS_EX_LAYOUTRTL*).

**Comment by rdipardo** [comment]
@donoh,
>  I confirm that only GDI (_WS_EX_LAYOUTRTL_, without _SCI_GETBIDIRECTIONAL_) is used in the current implementation of Notepad++, and it works without enabling _SC_TECHNOLOGY_DIRECTWRITE_. 

By "works" you probably mean "does what @Yaron10 listed as functional requirements with minimal interference". In fact, even N++'s bespoke implementation of v4.4.6 wrongly inverts the order of non-alphanumeric characters, as Geany [reportedly does][0] given Scintilla 5.1.5 and a platform where GDI does not exist and RTL has never been supported.

![npp_833_rtl](https://user-images.githubusercontent.com/59004801/162587349-ce88c8c4-6fb8-4016-9042-dba791331341.gif)

The recipe for RTL support in N++ is most likely too platform-dependent to justify upstream revisions. (The Geany bug did lead to [a minor patch][1], but only to avoid application crashes).

[0]: https://sourceforge.net/p/scintilla/bugs/2309
[1]: https://sourceforge.net/p/scintilla/code/ci/0759b3980fda03919b33d7be85a1653d27b8437f

**Comment by Yaron10** [comment]
@donho,

Thank you for the clarification. 👍 

@rdipardo,

> N++'s bespoke implementation of v4.4.6 wrongly inverts the order of non-alphanumeric characters...

I've never used Geany, but this is the behavior in the best-RTL-support editors.

Thank you. 👍

**Comment by rdipardo** [comment]
> I've never used Geany . . .

<https://www.geany.org/download/releases>

It's indispensable on Linux, where the package maintainers usually insulate you from the universe of platform quirks. The Windows version is just an impoverished simulacrum with no [virtual terminal][0].

[0]: https://github.com/geany/geany/issues/3145#issuecomment-1086852775

**Comment by Yaron10** [comment]
@rdipardo,

Thanks for that too.

The LTR/RTL behavior is notoriously complex. 
I thought it was better not to add non-alphanumeric characters to this issue. :)
But, reading your post again, the patch you mentioned might turn out as important.

**Comment by nyamatongwe** [comment]
> @donho: That makes us think the text direction management might have been changed in Scintilla5 with Microsoft legacy GDI (by setting WS_EX_LAYOUTRTL).

OK, the problem occurs with buffered drawing turned on. It happens there because the `LAYOUT_RTL` HDC state caused by `WS_EX_LAYOUTRTL` wasn't copied to the HDC of the bitmap used for buffering and that occurred because changeset [8a70f6](https://sourceforge.net/p/scintilla/code/ci/8a70f6af91ddd3c5e24ec2aad92b6c98e9363d9b/) switched to a different API to create the bitmap.

May be fixed with commit [fa80f5](https://sourceforge.net/p/scintilla/code/ci/fa80f5d42d95cb95030f587b00174471d946353b/) so update from the hg repository.

**Comment by Yaron10** [comment]
@nyamatongwe,

Fixed! 👍 

Thank you very much for the quick fix. Highly appreciated.

It seems I completely misunderstood your question about inverted glyphs.

@donho,

Thank you very much for looking into it and pushing. Highly appreciated.

@rdipardo,

Thank you very much for looking into it and testing. Highly appreciated.

**Comment by Yaron10** [comment]
Two points to complete this thread:

An interesting comment regarding [RTL and DirectWrite](https://sourceforge.net/p/scintilla/feature-requests/1435/#d7f5).

> Apart from the old **selection issue**, everything else seems to work perfectly.

Paste `I want to mention`.
Change to RTL.
Double-click "mention".

**Result:**
Random characters are selected.

**Note:** 
Double-clicking a Hebrew word in **RTL**, the word is selected properly.
Double-clicking a Hebrew word in **LTR**, random characters are selected.

Thank you all again.

**Comment by donho** [comment]
@nyamatongwe Thank you Neil for this quick fix!

@Yaron10 
> Paste I want to mention.
Change to RTL.
Double-click "mention".

> Result:
Random characters are selected.

I just tested on the fixed version and compared the result with v8.3.3 - the results are the same.
So it could be a point to be enhanced, but not a regression to me.

**Comment by Yaron10** [comment]
@donho,

> Apart from the old selection issue, everything else seems to work perfectly.

I used that sentence at the beginning of this thread.
To complete this thread I just wanted to explain it.

And you're right: this is an old issue.

Thank you once again.
